dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/message/ProgramMessageOperationParamMapper.java

```diff
@@ -0,0 +1,114 @@
+import org.hisp.dhis.common.BaseIdentifiableObject;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.common.IllegalQueryException;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.feedback.NotFoundException;
+import org.hisp.dhis.program.Enrollment;
+import org.hisp.dhis.program.Event;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramService;
+import org.hisp.dhis.user.CurrentUserUtil;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserService;
+import org.hisp.dhis.util.ObjectUtils;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Zubair Asghar
+ */
+@Component
+@RequiredArgsConstructor
+public class ProgramMessageOperationParamMapper {
+  private final IdentifiableObjectManager manager;
+
+  private final ProgramService programService;
+
+  private final UserService userService;
+
+  @Transactional(readOnly = true)
+  public ProgramMessageQueryParams map(ProgramMessageOperationParams operationParams)
+      throws NotFoundException {
+    Enrollment enrollment = getEntity(operationParams.getEnrollment(), Enrollment.class);
+    Event event = getEntity(operationParams.getEvent(), Event.class);
+
+    currentUserHasAccess(enrollment, event);
+
+    return ProgramMessageQueryParams.builder()
+        .enrollment(enrollment)
+        .event(event)
+        .afterDate(operationParams.getAfterDate())
+        .messageStatus(operationParams.getMessageStatus())
+        .beforeDate(operationParams.getBeforeDate())
+        .page(operationParams.getPage())
+        .pageSize(operationParams.getPageSize())
+        .organisationUnit(operationParams.getOu())
+        .build();
+  }
+
+  private <T extends BaseIdentifiableObject> T getEntity(UID entity, Class<T> klass)
+      throws NotFoundException {
+    if (entity == null) {
+      return null;
+    }
+
+    return Optional.ofNullable(manager.get(klass, entity.getValue()))
+        .orElseThrow(
+            () ->
+                new NotFoundException(
+                    String.format(
+                        "%s: %s does not exist.", klass.getSimpleName(), entity.getValue())));
+  }
+
+  private void currentUserHasAccess(Enrollment enrollment, Event event) {
+    Enrollment entity =
+        ObjectUtils.firstNonNull(
+            enrollment, Optional.ofNullable(event).map(Event::getEnrollment).orElse(null));
+
+    if (entity == null) {
+      throw new IllegalQueryException("Enrollment or Event has to be provided");
+    }
+
+    List<Program> programs = programService.getCurrentUserPrograms();
+
+    User currentUser = userService.getUserByUsername(CurrentUserUtil.getCurrentUsername());
+    if (currentUser != null && !programs.contains(entity.getProgram())) {
+      throw new IllegalQueryException(
+          String.format(
+              "User:%s does not have access to the required program:%s",
```

Comment:
if the `currentUser` is null, should we provide a more accurate error message?
in this case it's not that the user has no access, it's that something went wrong and we couldn't retrieve the user, right?

Reply:
is it possible to have `currentUser` as null? 
Even in async mode there will be `systemUser` available.

Reply:
I guess this is a question for @netroms